### PR TITLE
feat: migrate podGroup controller to ctrl runtime

### PR DESCRIPTION
**Pull request by zwpaper**
Signed-off-by: Wei Zhang <kweizh@gmail.com>

#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #
https://github.com/kubernetes-sigs/scheduler-plugins/issues/485

#### Special notes for your reviewer:

this is not fully tested, but please take a look to make sure we are on the same page
@Huang-Wei 

#### Does this PR introduce a user-facing change?

```release-note
migrate PodGroup to use controller-runtime controllers
```

### CODE REVIEW

File: pkg/controllers/coscheduling/podgroup_controller.go [hunk 1]
```diff
@@ -47,16 +66,153 @@ type PodGroupReconciler struct {
...
+	}
+
+	pgCopy := pg.DeepCopy()
+	podList := &corev1.PodList{}
+	err = r.List(ctx, podList,
+		client.MatchingLabelsSelector{
+			Selector: labels.Set(map[string]string{
+				schedv1alpha1.PodGroupLabel: pgCopy.Name}).AsSelector(),
+		})
+	if err != nil {
```
Comment (PiotrProkop) [inline pkg/controllers/coscheduling/podgroup_controller.go:99]:
```suggestion
        labelSelector := client.MatchingLabelsSelector{
			Selector: labels.Set(map[string]string{
				schedv1alpha1.PodGroupLabel: pg.Name}).AsSelector(),
		}
	if err = r.List(ctx, podList, labelSelector); err != nil {
```

File: pkg/controllers/coscheduling/podgroup_controller.go [hunk 2]
```diff
@@ -47,16 +66,153 @@ type PodGroupReconciler struct {
...
+			for _, pod := range pods {
+				switch pod.Status.Phase {
+				case v1.PodRunning:
+					running++
+				case v1.PodSucceeded:
+					succeeded++
+				case v1.PodFailed:
+					failed++
+				}
+			}
```
Comment (PiotrProkop) [inline pkg/controllers/coscheduling/podgroup_controller.go:129]: put this into a function that returns 3 values, something like or similliar:
```go
func getCurrentPodStats(podList v1.PodList) (int32,int32,int32){
}
```

File: pkg/controllers/coscheduling/podgroup_controller.go [hunk 3]
```diff
@@ -47,16 +66,153 @@ type PodGroupReconciler struct {
...
+		if pgCopy.Status.Succeeded >= pg.Spec.MinMember {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupFinished
+		}
+	}
+
+	err = r.patchPodGroup(ctx, pg, pgCopy)
+	return ctrl.Result{}, err
+}
+
+func (r *PodGroupReconciler) patchPodGroup(ctx context.Context, old, new *schedv1alpha1.PodGroup) error {
```
Comment (PiotrProkop) [inline pkg/controllers/coscheduling/podgroup_controller.go:161]: make this return `ctrl.Result, error` so you can just `return r.patchPodGroup` in `Reconcile`.

File: pkg/controllers/coscheduling/podgroup_controller.go [hunk 4]
```diff
@@ -47,16 +66,153 @@ type PodGroupReconciler struct {
...
+	}
+	return nil
+}
+
+func fillOccupiedObj(pg *schedv1alpha1.PodGroup, pod *corev1.Pod) {
+	var refs []string
+	for _, ownerRef := range pod.OwnerReferences {
+		refs = append(refs, fmt.Sprintf("%s/%s", pod.Namespace, ownerRef.Name))
+	}
+	if len(pg.Status.OccupiedBy) == 0 {
```
Comment (PiotrProkop) [inline pkg/controllers/coscheduling/podgroup_controller.go:180]: this check should be first in a func, not to perform unnecessary loop if this returns early

File: pkg/controllers/coscheduling/podgroup_controller.go [hunk 5]
```diff
@@ -47,16 +66,153 @@ type PodGroupReconciler struct {
...
+	}
+	if len(refs) != 0 {
+		sort.Strings(refs)
+		pg.Status.OccupiedBy = strings.Join(refs, ",")
+	}
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *PodGroupReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.recorder = mgr.GetEventRecorderFor("PodGroupController")
```
Comment (PiotrProkop) [inline pkg/controllers/coscheduling/podgroup_controller.go:191]: this recorder is unused you can just not initiate it.

File: pkg/controllers/coscheduling/podgroup_controller.go [hunk 6]
```diff
@@ -18,17 +18,36 @@ package controllers
...
 
-	v1alpha1 "sigs.k8s.io/scheduler-plugins/apis/scheduling/v1alpha1"
+	schedv1alpha1 "sigs.k8s.io/scheduler-plugins/apis/scheduling/v1alpha1"
+	"sigs.k8s.io/scheduler-plugins/pkg/util"
 )
 
 // PodGroupReconciler reconciles a PodGroup object
 type PodGroupReconciler struct {
+	log      logr.Logger
+	recorder record.EventRecorder
```
Comment (PiotrProkop) [inline pkg/controllers/coscheduling/podgroup_controller.go:49]: this is unused you can remove it.

File: cmd/controller/app/server.go [hunk 7]
```diff
@@ -20,22 +20,42 @@ import (
...
 )
 
+var (
+	scheme   = runtime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+
```
Comment (PiotrProkop) [inline cmd/controller/app/server.go:55]: remove newline

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 6 above)
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_controller.go:49]: `recorder` will be used in later development, and it is worthwhile to add it here when we first migrate to controller runtime in this controller runtime way

File: pkg/controllers/coscheduling/podgroup_controller.go [hunk 8]
```diff
@@ -47,16 +66,153 @@ type PodGroupReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
 func (r *PodGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	log := log.FromContext(ctx)
+	pg := &schedv1alpha1.PodGroup{}
+	err := r.Get(ctx, req.NamespacedName, pg)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_controller.go:72]: the `err` will be used later, it has to create a new one if wrapped inside if

File: pkg/controllers/coscheduling/podgroup_controller.go [hunk 9]
```diff
@@ -47,16 +66,153 @@ type PodGroupReconciler struct {
...
+		if apierrs.IsNotFound(err) {
+			log.V(5).Info("Pod group has been deleted")
+			return ctrl.Result{}, nil
+		}
+		log.V(3).Error(err, "Unable to retrieve pod group")
+		return ctrl.Result{}, err
+	}
+
+	if pg.Status.Phase == schedv1alpha1.PodGroupFinished ||
```
[the remark this answers is not in the document]
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_controller.go:81]: we are working on different conditions here, and doing an early return,

separating the different conditions makes more sense to me,

1. check phase, if nothing to do, early return
2. check scheduled status, if nothing to do, early return

it looks more chaotic if we mixed them up. straight early returning seems better than wrapping things inside a big switch.

btw, we have not yet read pods, so we can not do the fillOccupiedObj in this big switch.

File: pkg/controllers/coscheduling/podgroup_controller.go [hunk 10]
```diff
@@ -47,16 +66,153 @@ type PodGroupReconciler struct {
...
+		return ctrl.Result{}, nil
+	}
+	// If startScheduleTime - createTime > 2days,
+	// do not enqueue again because pod may have been GCed
+	if pg.Status.Scheduled == pg.Spec.MinMember && pg.Status.Running == 0 &&
+		pg.Status.ScheduleStartTime.Sub(pg.CreationTimestamp.Time) > 48*time.Hour {
+		return ctrl.Result{}, nil
+	}
+
+	pgCopy := pg.DeepCopy()
```
[the remark this answers is not in the document]
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_controller.go:92]: we should do the copy when all precheck passes, and will begin to do the real logic, right?

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 1 above)
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_controller.go:99]: the `labelSelector` will only be used once in this `List`, why create a tmp var for it?

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 2 above)
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_controller.go:129]: make sense, will do it later

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 3 above)
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_controller.go:161]: actually, I may put the `patch` inside a `defer` earlier, will change this later

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 4 above)
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_controller.go:180]: make sense

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 5 above)
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_controller.go:191]: it will be used later, initiating this when first migrated to controller runtime and we can use it directly if need to

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply pkg/controllers/coscheduling/podgroup_controller.go:92]: yes

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply pkg/controllers/coscheduling/podgroup_controller.go:81]: Then you should sort cases, so the one that returns early are on top, and wrap reading pods into a function and use it only in phases which requires them to do calculation. Right now it is hard to follow the logic with those long if condition statements.

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 1 above)
Comment (PiotrProkop) [reply pkg/controllers/coscheduling/podgroup_controller.go:99]: Yeah, you're right it's better to make func for that to make this more concise :
```go
func genLabelSelectorFor(name string) client.MatchingLabelsSelector{
    return client.MatchingLabelsSelector{
			Selector: labels.Set(map[string]string{
				schedv1alpha1.PodGroupLabel: name}).AsSelector(),
    }
}
...
if err = r.List(ctx, podList, genLabelSelectorFor(pg.Name)); err != nil {
```

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 5 above)
Comment (PiotrProkop) [reply pkg/controllers/coscheduling/podgroup_controller.go:191]: Than you can add it later, when you will use it.

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (PiotrProkop) [reply pkg/controllers/coscheduling/podgroup_controller.go:72]: yeah, this is common pattern how to check errors in Golang.

**Comment by Huang-Wei** [comment]
@denkensk do you have some cycles to review this PR? Given we need to manipulate different CRDs, I want to migrate to use controller-runtime (instead of typed clientset) to in both controllers and scheduler plugins.

This PR is about controllers.

**Comment by zwpaper** [comment]
hi @Huang-Wei @denkensk, I migrated the podGroup to ctrl runtime too, it should pass all the original podGroup tests, could you please add an `ok-to-test`, so I could improve the testing errors while I am testing it locally?

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 6 above)
Comment (denkensk) [reply pkg/controllers/coscheduling/podgroup_controller.go:49]: You can add TODO here.

File: pkg/controllers/coscheduling/podgroup_controller.go [hunk 11]
```diff
@@ -47,16 +66,158 @@ type PodGroupReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
 func (r *PodGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	log := log.FromContext(ctx)
+	pg := &schedv1alpha1.PodGroup{}
+	if err := r.Get(ctx, req.NamespacedName, pg); err != nil {
+		if apierrs.IsNotFound(err) {
+			log.V(5).Info("Pod group has been deleted")
```
Comment (denkensk) [inline pkg/controllers/coscheduling/podgroup_controller.go:73]: Need a print namespace and name of this PodGroup

File: pkg/controllers/coscheduling/podgroup_controller.go [hunk 12]
```diff
@@ -47,16 +66,158 @@ type PodGroupReconciler struct {
...
+	}
+
+	pgCopy := pg.DeepCopy()
+	podList := &corev1.PodList{}
+	if err := r.List(ctx, podList,
+		client.MatchingLabelsSelector{
+			Selector: labels.Set(map[string]string{
+				schedv1alpha1.PodGroupLabel: pgCopy.Name}).AsSelector(),
+		}); err != nil {
+		log.Error(err, "List pods for group failed")
```
Comment (denkensk) [inline pkg/controllers/coscheduling/podgroup_controller.go:100]: also need a print namespace and name of this PodGroup

File: pkg/controllers/coscheduling/podgroup_test.go [hunk 13]
```diff
@@ -214,47 +224,56 @@ func TestFillGroupStatusOccupied(t *testing.T) {
...
 	}
 }
 
-func setUp(ctx context.Context, podNames []string, pgName string, podPhase v1.PodPhase, minMember int32, groupPhase v1alpha1.PodGroupPhase, podGroupCreateTime *metav1.Time, podOwnerReference []metav1.OwnerReference) (*PodGroupController, *fake.Clientset, *pgfake.Clientset) {
-	var kubeClient *fake.Clientset
-	if len(podNames) == 0 {
-		kubeClient = fake.NewSimpleClientset()
-	} else {
+func setUp(ctx context.Context,
+	podNames []string, pgName string, podPhase v1.PodPhase, minMember int32,
```
Comment (denkensk) [inline pkg/controllers/coscheduling/podgroup_test.go:255]: need to format this argument

**Review by denkensk — COMMENTED** [review]
Thanks. @zwpaper

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 11 above)
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_controller.go:73]: the logger from controller-runtime context is printing namespace and name by default

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 12 above)
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_controller.go:100]: the logger from controller-runtime context is printing namespace and name by default

**Comment by zwpaper** [comment]
this is what the logs from controll-runtime klogr looks like:

```console
# kubectl logs -n scheduler-plugins scheduler-plugins-controller-6999f68449-xc6mz
W0214 17:10:19.175586       1 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0214 17:10:19.328949       1 logr.go:261] controller-runtime/metrics "msg"="Metrics server is starting to listen" "addr"=":8080"
I0214 17:10:19.329215       1 logr.go:261] setup "msg"="starting manager"
I0214 17:10:19.329310       1 elasticquota.go:115] "Starting Elastic Quota control loop"
I0214 17:10:19.329334       1 elasticquota.go:117] "Waiting for informer caches to sync"
I0214 17:10:19.329430       1 internal.go:362]  "msg"="Starting server" "addr"={"IP":"::","Port":8080,"Zone":""} "kind"="metrics" "path"="/metrics"
I0214 17:10:19.329470       1 internal.go:362]  "msg"="Starting server" "addr"={"IP":"::","Port":8081,"Zone":""} "kind"="health probe"
I0214 17:10:19.329568       1 controller.go:185]  "msg"="Starting EventSource" "controller"="podgroup" "controllerGroup"="scheduling.sigs.k8s.io" "controllerKind"="PodGroup" "source"="kind source: *v1alpha1.PodGroup"
I0214 17:10:19.329581       1 controller.go:185]  "msg"="Starting EventSource" "controller"="podgroup" "controllerGroup"="scheduling.sigs.k8s.io" "controllerKind"="PodGroup" "source"="kind source: *v1.Pod"
I0214 17:10:19.329586       1 controller.go:193]  "msg"="Starting Controller" "controller"="podgroup" "controllerGroup"="scheduling.sigs.k8s.io" "controllerKind"="PodGroup"
I0214 17:10:19.430002       1 elasticquota.go:122] "Elastic Quota sync finished"
I0214 17:10:19.430037       1 controller.go:227]  "msg"="Starting workers" "controller"="podgroup" "controllerGroup"="scheduling.sigs.k8s.io" "controllerKind"="PodGroup" "worker count"=1
I0214 17:10:19.430181       1 podgroup_controller.go:70]  "msg"="reconciling" "controller"="podgroup" "controllerGroup"="scheduling.sigs.k8s.io" "controllerKind"="PodGroup" "name"="pg1" "namespace"="default" "podGroup"={"name":"pg1","namespace":"default"} "reconcileID"="74db25c9-2a0a-45cd-9e6f-535e6bcb2384"
```

File: pkg/controllers/coscheduling/podgroup_test.go (hunk 13 above)
Comment (zwpaper) [reply pkg/controllers/coscheduling/podgroup_test.go]: @denkensk setUp uses the cases' options as args, there is no good way to reduce the number of args, maybe we can pass the case struct into setUp directly?

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 11 above)
Comment (denkensk) [reply pkg/controllers/coscheduling/podgroup_controller.go]: OK

File: pkg/controllers/coscheduling/podgroup_controller.go (hunk 12 above)
Comment (denkensk) [reply pkg/controllers/coscheduling/podgroup_controller.go]: OK

File: pkg/controllers/coscheduling/podgroup_test.go (hunk 13 above)
Comment (denkensk) [reply pkg/controllers/coscheduling/podgroup_test.go]: Maybe we put an arg per line here to make it looks clear?

**Comment by zwpaper** [comment]
hi @denkensk, PR updated, PTAL

File: pkg/controllers/podgroup_test.go [hunk 14]
```diff
@@ -214,47 +224,60 @@ func TestFillGroupStatusOccupied(t *testing.T) {
...
+	podGroupCreateTime *metav1.Time,
+	podOwnerReference []metav1.OwnerReference) (*PodGroupReconciler, client.WithWatch) {
+	s := scheme.Scheme
+	pg := makePG(pgName, minMember, groupPhase, podGroupCreateTime)
+	s.AddKnownTypes(schedulingv1a1.SchemeGroupVersion, pg)
+	objs := []runtime.Object{pg}
+	if len(podNames) != 0 {
 		ps := makePods(podNames, pgName, podPhase, podOwnerReference)
-		kubeClient = fake.NewSimpleClientset(ps[0], ps[1])
+		// s.AddKnownTypes(clientgoscheme.SchemeGroupVersion, ps)
```
Comment (denkensk) [inline pkg/controllers/podgroup_test.go:268]: remove this line.

**Review by denkensk — COMMENTED** [review]
Thanks. Looks to me overall. Some small comments. @zwpaper

**Comment by zwpaper** [comment]
Hi @denkensk, PR updated again, PTAL, thanks

**Comment by denkensk** [comment]
/lgtm
/approve

Thanks for your contribution @zwpaper
